src/sasl/cyrsasl_scram.erl

```diff
@@ -58,164 +58,196 @@ mech_new(_Host, Creds) ->
-                                                                                  mongoose_scram:iterations()),
-                                                   {mongoose_scram:stored_key(mongoose_scram:client_key(SaltedPassword)),
-                                                    mongoose_scram:server_key(SaltedPassword),
-                                                    TempSalt,
-                                                    mongoose_scram:iterations()}
-                                            end,
-                                            {NStart, _} = binary:match(ClientIn, <<"n=">>),
-                                            ClientFirstMessageBare = binary:part(ClientIn,
-                                                                                 {NStart,
-                                                                                  byte_size(ClientIn)-NStart}),
-                                            ServerNonce =
-                                            jlib:encode_base64(crypto:strong_rand_bytes(?NONCE_LENGTH)),
-                                            ServerFirstMessage =
-                                            iolist_to_binary(
-                                              [<<"r=">>,
-                                               ClientNonce,
-                                               ServerNonce,
-                                               <<",s=">>,
-                                               jlib:encode_base64(Salt),
-                                               <<",i=">>,
-                                               integer_to_list(IterationCount)]),
-                                            {continue, ServerFirstMessage,
-                                             State#state{step = 4, stored_key = StoredKey,
-                                                         server_key = ServerKey,
-                                                         auth_message =
-                                                         <<ClientFirstMessageBare/binary,
-                                                           ",", ServerFirstMessage/binary>>,
-                                                         client_nonce = ClientNonce,
-                                                         server_nonce = ServerNonce,
-                                                         username = UserName,
-                                                         auth_module = AuthModule}}
-                                    end;
-                                _Else -> {error, <<"not-supported">>}
-                            end
-                    end
+    try
+        % Check received attributes
+        ClientInList = re:split(ClientIn, <<",">>, [{return, binary}]),
+        ok = check_extension_attribute(ClientInList),
+        [CBind, _AuthIdentity, UserNameAtt, ClientNonceAtt | _] = ClientInList,
+        ok = check_message_headers(CBind),
```

Comment:
For example, these three lines could be unified as
```erlang
        {UserNameAtt, ClientNonceAtt} = check_extension_attribute(ClientInList),
```
if `check_extension_attribute/1` was implemented as
```erlang
check_extension_attribute([_,_,_,_, Extension | _]) when Extension /= [] ->
    erlang:error(<<"protocol-error-extension-not-supported">>);
check_extension_attribute([CBind | _])
  when (CBind =/= <<"y">>) andalso (CBind =/= <<"n">>) ->
    erlang:error(<<"bad-protocol">>);
check_extension_attribute([_CBind, _AuthIdentity, UserNameAtt, ClientNonceAtt | _]) ->
    {UserNameAtt, ClientNonceAtt}.
```
But I'm being very opinionated here, just giving ideas 😄